planner/core/collect_column_stats_usage.go

```diff
@@ -133,119 +152,153 @@ func (p *LogicalUnionAll) updateColMapAndAddPredicateColumns(c *predicateColumnC
 	}
 }
 
-func (c *predicateColumnCollector) collectFromPlan(lp LogicalPlan) {
+func (ds *DataSource) addHistNeededColumns(c *columnStatsUsageCollector) {
+	tblID := ds.TableInfo().ID
```

Comment:
Should it be physicalTableID?

Reply:
Seems physicalTableID is not set when it's non-partition table.
https://github.com/pingcap/tidb/pull/30754/files#diff-3b47fd3bf2c33ae3a03037a65ed39df33a4ae73debe9cf6f5bb2f397508061a0R4099

Reply:
Yes, thinks for reminding. We can first check `DataSource.isPartition`. If it is true, use `DataSource.physicalTableID`. Otherwise use `DataSource.TableInfo().ID`.

Reply:
I see, after #30754 is merged we can just use `physicalTableID`.

Reply:
Fixed it and added test.